core/src/main/java/com/linecorp/armeria/client/endpoint/StickyStrategy.java

```diff
@@ -0,0 +1,140 @@
+ * user passed {@code RequestContextHasher} with consistent hashing.
+ */
+public class StickyStrategy implements EndpointSelectionStrategy {
+
+    private final RequestContextHasher requestHasher;
+
+    private final EndpointHasher endpointHasher;
+
+    private static final EndpointHasher DEFAULT_ENDPOINT_HASHER = Endpoint::hashCode;
+
+    /**
+     * An interface represents strategy to hash a ClientRequestContext to a long.
+     */
+    @FunctionalInterface
+    public interface RequestContextHasher {
+        /**
+         * Hash an {@code ClientRequestContext} to a Long.
+         * @param ctx The {@code ClientRequestContext} request context
+         * @return A key related to current request, which used to load balance
+         */
+        long hashOf(ClientRequestContext ctx);
+    }
+
+    /**
+     * An interface represents strategy to hash a ClientRequestContext to a long.
+     */
+    @FunctionalInterface
+    public interface EndpointHasher {
+        /**
+         * Hash an {@code Endpoint} to a Long.
+         * @param endpoint The {@code Endpoint}
+         * @return A key related to the {@code endpoint}, which used to load balance
+         */
+        long hashOf(Endpoint endpoint);
+    }
+
+    /**
+     * Return new StickyStrategy.
+     * @param requestHasher A {@code RequestContextHasher}
+     */
+    public StickyStrategy(RequestContextHasher requestHasher) {
+        this.requestHasher = requestHasher;
+        this.endpointHasher = DEFAULT_ENDPOINT_HASHER;
+    }
+
+    /**
+     * Return new StickyStrategy.
+     * @param requestHasher A {@code RequestContextHasher}
+     * @param endpointHasher A {@code EndpointHasher}
+     */
+    public StickyStrategy(RequestContextHasher requestHasher,
+                          EndpointHasher endpointHasher) {
+        this.requestHasher = requestHasher;
+        this.endpointHasher = endpointHasher;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new StickySelector(requestHasher, endpointGroup, this);
+    }
+
+    private static final class StickySelector implements EndpointSelector {
+
+        private final RequestContextHasher hasher;
+        private final EndpointGroup endpointGroup;
+        private final EndpointSelectionStrategy endpointSelectionStrategy;
+        private static final TreeMap<Long, Endpoint> buckets = new TreeMap<>();
+
+        StickySelector(RequestContextHasher hasher, EndpointGroup endpointGroup,
+                       EndpointSelectionStrategy endpointSelectionStrategy) {
+            this.hasher = hasher;
+            this.endpointGroup = endpointGroup;
+            this.endpointSelectionStrategy = endpointSelectionStrategy;
+            if (endpointGroup instanceof DynamicEndpointGroup) {
+                DynamicEndpointGroup dynamicEndpointGroup = (DynamicEndpointGroup) endpointGroup;
```

Comment:
Don't need to cast to DynamicEndpointGroup since all EndpointGroup implement Listenable